src/sleep/sleep.c

```diff
@@ -243,32 +232,37 @@ static int execute_s2h(usec_t hibernate_delay_sec) {
         if (r < 0)
                 return r;
 
-        wake_time = original_time + DIV_ROUND_UP(hibernate_delay_sec, USEC_PER_SEC);
-        r = rtc_write_wake_alarm(wake_time);
-        if (r < 0)
-                return r;
+        tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
+        if (tfd < 0)
+                return log_error_errno(tfd, "Error creating timerfd");
 
-        log_debug("Set RTC wake alarm for %" PRIu64, wake_time);
+        wake_time = DIV_ROUND_UP(hibernate_delay_sec, USEC_PER_SEC);
+
+        log_debug("Set timerfd wake alarm for %"PRIu64 " seconds", wake_time);
+        ts.it_value.tv_sec = wake_time;
+        r = timerfd_settime(tfd, 0, &ts, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Error setting hibernate timer");
 
         r = execute(suspend_modes, suspend_states);
         if (r < 0)
                 return r;
 
-        /* Reset RTC right-away */
-        r = rtc_write_wake_alarm(0);
+        r = close(tfd);
         if (r < 0)
-                return r;
+                return log_error_errno(r, "Error closing timerfd");
 
         r = rtc_read_time(&cmp_time);
         if (r < 0)
                 return r;
 
         log_debug("Woke up at %"PRIu64, cmp_time);
 
-        if (cmp_time < wake_time) /* We woke up before the alarm time, we are done. */
+        if (cmp_time < (original_time + wake_time)) /* We woke up before the alarm time, we are done. */
```

Comment:
instead of checking this by comparing times, please instead use poll() to check if the timerfd is readable after we come back from suspend. That's how the kernel tells us whether the time out hit or not, and we can rely on that.